### ISSUE TITLE
Add Pixelated Butterfly for efficient sparse training

### ISSUE THREAD

**Issue by lewtun** [huggingface/optimum#56]
## Description

<img width="1025" alt="Screen Shot 2021-12-12 at 17 04 51" src="https://user-images.githubusercontent.com/26859204/145719779-c17f49f2-eb8b-48f5-8126-5ff61058e4f1.png">

Pixelated butterfly (Pixelfly) is a training technique that use a simple fixed sparsity pattern based on flat block butterfly and low-rank matrices to sparsify most network layers. From the paper:

> On the ImageNet classification and WikiText-103 language modeling tasks, our sparse models train up to 2.5× faster than the dense MLP-Mixer, Vision Transformer, and GPT-2 medium with no drop in accuracy.

This seems like an interesting approach to consider for integration with `optimum`, and builds on prior work by @madlag using [`pytorch_block_sparse`](https://github.com/huggingface/pytorch_block_sparse).

* Paper: https://arxiv.org/abs/2112.00029
* Twitter thread: https://twitter.com/BeidiChen/status/1469135402850082816?s=20
